File: clients/typescript/examples/basic.ts
import { Client } from "../src/index.js";

async function main() {
  const [, , binary, harness] = process.argv;
  if (!binary || !harness) {
    console.error("usage: tsx examples/basic.ts <binary_path> <harness>");
    process.exit(2);
  }

  const client = new Client("http://127.0.0.1:8080");
  const conv = await client.open({ harness, binaryPath: binary });
  console.log(`opened conversation ${conv.id}`);

  try {
    await conv.withControl(async () => {
      const turnId = await conv.send("hello");
      console.log(`sent turn ${turnId}`);
      for await (const ev of conv.events()) {
        console.log(`  event: turn=${ev.turn.id} state=${ev.turn.state}`);
        if (ev.turn.id === turnId && (ev.turn.state === "complete" || ev.turn.state === "errored")) {
          break;
        }
      }
    });
    for (const t of await conv.history()) {
      console.log(`  ${t.role}: ${JSON.stringify((t.text ?? "").slice(0, 80))}`);
    }
  } finally {
    await conv.close();
  }
}

try {
  await main();
} catch (e) {
  console.error(e);
  process.exit(1);
}

File: clients/python/examples/basic.py
"""End-to-end smoke test against a running harness-chatd.

Usage:
    cd clients/python && PYTHONPATH=. python3 examples/basic.py /usr/local/bin/codex codex
"""

import sys
from harness_chat import Client


def main() -> None:
    if len(sys.argv) != 3:
        print("usage: basic.py <binary_path> <harness>", file=sys.stderr)
        sys.exit(2)
    binary, harness = sys.argv[1], sys.argv[2]

    client = Client("http://127.0.0.1:8080")
    conv = client.open(harness=harness, binary_path=binary)
    print(f"opened conversation {conv.id}")
    try:
        with conv.control():
            turn_id = conv.send("hello")
            print(f"sent turn {turn_id}")
            for ev in conv.events():
                if ev.type != "turn" or ev.turn is None:
                    continue
                print(f"  event: turn={ev.turn.id} state={ev.turn.state}")
                if ev.turn.id == turn_id and ev.turn.state in ("complete", "errored"):
                    break
        for t in conv.history():
            print(f"  {t.role}: {t.text[:80]!r}")
    finally:
        conv.close()


if __name__ == "__main__":
    main()
